include/circt/Dialect/SMT/SMTOps.td

```diff
@@ -337,4 +338,64 @@ def ImpliesOp : SMTOp<"implies", [Pure]> {
   let assemblyFormat = "$lhs `,` $rhs attr-dict";
 }
 
+class QuantifierOp<string mnemonic> : SMTOp<mnemonic, [
+  RecursivelySpeculatable,
+  RecursiveMemoryEffects,
+  SingleBlockImplicitTerminator<"smt::YieldOp">,
+]> {
+  let description = [{
+    This operation represents the }] # summary # [{ as described in the
+    [SMT-LIB 2.6 standard](https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2021-05-12.pdf).
+    It is part of the language itself rather than a theory or logic.
+
+    The operation specifies the names (as an attribute) and types (as the types
+    of the block arguments of the regions) of bound variables that may be
+    used in the 'body' of the operation. If a 'patterns' region is specified,
+    the block arguments must match the ones of the 'body' region and (other
+    than there) must be used at least once in the 'patterns' region. It may also
+    not contain any operations that bind variables, such as quantifiers. While
+    the 'body' region must always yield exactly one `!smt.bool`-typed value, the
+    'patterns' region can yield an arbitrary number (but at least one) of SMT
+    values.
+
+    The 'no_patterns' attribute is only allowed when no 'patterns' region is
+    specified and forbids the solver to generate and use patterns for this
+    quantifier.
+
+    The 'weight' attribute indicates the importance of this quantifier being
+    instantiated compared to other quantifiers that may be present. The default
+    value is zero.
```

Comment:
I couldn't find information about that attribute on the SMT-LIB documentation, do you have a link to it?

Reply:
SMT-LIB allows to annotate most (all?) terms with attributes (see SMT-LIB 2.6, section 3.4 and 3.6 (specifically the grammar figure in the beginning and 3.6.5)). Those annotations are not allowed to change the semantics of the language but allow adding metadata for solvers, such as this `weight` attribute which is supported, e.g., by Z3: https://stackoverflow.com/questions/62512506/which-smt-lib-attributes-does-z3-support-and-why

The `no-pattern` attribute is another one. Apparently, only a few attributes are important enough to be mentioned in the standard (such as `:pattern`).

My plan is to support (at least some of) the attributes supported by solvers but not mentioned in the standard such that the passes that generate SMT IR can have more control over the solver.

I can modify the documentation here to make this a bit clearer.

Reply:
That's perfect thanks!